source/server/drain_manager_impl.cc

```diff
@@ -35,30 +32,24 @@ bool DrainManagerImpl::drainClose() const {
     return false;
   }
 
-  // We use the tick time as in increasing chance that we shutdown connections.
-  return static_cast<uint64_t>(drain_time_completed_.load()) >
-         (server_.random().random() % server_.options().drainTime().count());
-}
-
-void DrainManagerImpl::drainSequenceTick() {
-  ENVOY_LOG(trace, "drain tick #{}", drain_time_completed_.load());
-  ASSERT(drain_time_completed_.load() < server_.options().drainTime().count());
-  ++drain_time_completed_;
+  const auto remaining_time = std::chrono::duration_cast<std::chrono::seconds>(
+      drain_deadline_ - server_.dispatcher().timeSource().monotonicTime());
+  const auto elapsed_time = server_.options().drainTime() - remaining_time;
 
-  if (drain_time_completed_.load() < server_.options().drainTime().count()) {
-    drain_tick_timer_->enableTimer(std::chrono::milliseconds(1000));
-  } else if (drain_sequence_completion_) {
-    drain_sequence_completion_();
-  }
+  // P(shut down connections) = elapsed_drain_time / overall_drain_timeout
+  return static_cast<uint64_t>(elapsed_time.count()) >
+         (server_.random().random() % server_.options().drainTime().count());
```

Comment:
note that this has performance implications in terms of asking for a new random number for each drain decision. I'm not sure that this matters but it's worth calling out. 

Stepping back, what is the reasoning for this change? It's not clear to me why this needs to change.

Reply:
I don't think the random() calls have changed, the existing code was already calling for a random() on each drain decision.

Reasoning for the change:

1. I'm planning to enhance integration testing coverage around the drain period given the additional complexity I'm planning to add. I'd like to reliably use simulated time to write those tests.

2. The current approach rearms the timer every second just for the purpose of updating the drain time completed, but this seems less precise since it has to run the code prior to the rearming on each callback. Having a single timer arming makes the behaviour easier to reason about in simulated time.

3. I found the existing code difficult to understand at a glance, and it wasn't sufficiently tested. Computing the deadline one time at the start simplifies the code a lot, I had a hard time understanding the whole drainSequenceTick thing. It's also easier to extend or modify, in case I end up needing to deal with overriding drain close periods or new phases in the drain sequence.

Reply:
OK I missed the diff in which we did random in the old code. OK fair enough sounds fine to me.